subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2020 InnBlue
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tftp_client, LOG_LEVEL_DBG);
+
+/* Public / Internal TFTP header file. */
+#include <net/tftp.h>
+#include "tftp_client.h"
+
+/* TFTP Global Buffer. */
+static u8_t   tftpc_buffer[TFTPC_MAX_BUF_SIZE];
+static u32_t  tftpc_buffer_size;
+
+/* TFTP Request block number property. */
+static u32_t  tftpc_block_no;
+static u32_t  tftpc_index;
+
+/* Global mutex to protect critical resources. */
+K_MUTEX_DEFINE(tftpc_lock);
+
+/* Name: make_request
+ * Description: This function takes in a given list of parameters and
+ * returns a read request packet. This packet can be sent
+ * out directly to the TFTP server.
+ */
+static u32_t make_request(const char *remote_file, const char *mode,
+			  u8_t request_type)
+{
+	u32_t      req_size;
+	const char def_mode[] = "octet";
+
+	/* Populate the read request with the provided params. Note that this
+	 * is created per RFC1350.
+	 */
+
+	/* Fill in the Request Type. Also keep tabs on the request size. */
+	sys_put_be16(request_type, tftpc_buffer);
+	req_size = 2;
+
+	/* Copy in the name of the remote file - the file we need to get
+	 * from the server. Add an upper bound to ensure no buffers overflow.
+	 */
+	strncpy(tftpc_buffer + req_size, remote_file, TFTP_MAX_FILENAME_SIZE);
+	req_size += strlen(remote_file);
+
+	/* Fill in 0. */
+	tftpc_buffer[req_size] = 0x0;
+	req_size++;
+
+	/* Default to "Octet" if mode not specified. */
+	if (mode == NULL) {
+		mode = def_mode;
+	}
+
+	/* Copy the mode of operation. */
+	strncpy(tftpc_buffer + req_size, mode, TFTP_MAX_MODE_SIZE);
+	req_size += strlen(mode);
+
+	/* Fill in 0. */
+	tftpc_buffer[req_size] = 0x0;
+	req_size++;
+
+	return req_size;
+}
+
+/* Name: send_err
+ * Description: This function sends an Error report to the TFTP Server.
+ */
+static inline int send_err(int sock, int err_code, char *err_string)
+{
+	u32_t req_size;
+
+	LOG_ERR("Client Error. Sending code: %d(%s)", err_code, err_string);
+
+	/* Fill in the "Err" Opcode and the actual error code. */
+	sys_put_be16(ERROR_OPCODE, tftpc_buffer);
+	sys_put_be16(err_code, tftpc_buffer + 2);
+	req_size = 4;
+
+	/* Copy the Error String. */
+	strcpy(tftpc_buffer + req_size, err_string);
+	req_size += strlen(err_string);
+
+	/* Send Error to server. */
+	return send(sock, tftpc_buffer, req_size, 0);
+}
+
+/* Name: tftpc_recv_data
+ * Description: This function tries to get data from the TFTP Server
+ * (either response or data). Times out eventually.
+ */
+static int tftpc_recv_data(int sock)
+{
+	int           stat;
+	struct pollfd fds;
+
+	/* Setup FDS. */
+	fds.fd     = sock;
+	fds.events = ZSOCK_POLLIN;
+
+	/* Enable poll on this socket. Wait for the specified number
+	 * of seconds before exiting the call.
+	 */
+	stat = poll(&fds, 1, K_SECONDS(CONFIG_TFTPC_REQUEST_TIMEOUT));
```

Comment:
This will fail after we start to use k_timeout_t because K_SECONDS() will return k_timeout_t instead of int. In order to fix that just change the timeout to `MSEC_PER_SEC * CONFIG_TFTPC_REQUEST_TIMEOUT`

Reply:
Fixed..

Reply:
> MSEC_PER_SEC * CONFIG_TFTPC_REQUEST_TIMEOUT

I'd suggest that having timeout in seconds is very coarse-grained, and would suggest to change it to be in milliseconds in the first place (figures like 0.5s and 1.5s are very valid for timeout).

Reply:
Yeah, I have to agree with this.. Pushed and tested.. Now the Kconfig option is "msec" instead of "sec"..

Reply:
> Now the Kconfig option is "msec" instead of "sec"..

Thanks!